### DISCUSSION TITLE
[Feature Request] Built-In Support for Sitemaps

### DISCUSSION THREAD

**Discussion by sakulstra** [vercel/next.js#26422] [Ideas]
Currently, there isn't a built-in way to understand every page generated by Next.js at build-time. **It would be great if Next.js supported a built-in way to generate a sitemap, which could be generated as part of the build.** 

There are now [official docs on sitemaps](https://github.com/vercel/next.js/pull/28346), but they still require custom user code.

**Comment by natemoo-re** [comment]
I would really love to see this integrated into Next. Just to note—If [[RFC] Plugins](https://github.com/zeit/next.js/issues/9133) lands, this could be solved in user land.

As far as DX goes, I would like to see a solution where `sitemap` builds on top of `getStaticProps` and `getServerSideProps`.

```js
export async function getStaticProps({ params }) {
  const response = await fetch(`https://example.com/post/${params.id}`)
  const post = await response.json();

  return {
    props: post,
    // see https://www.sitemaps.org/protocol.html#xmlTagDefinitions
    sitemap: {
     lastmod: new Date('2020-05-06'), // optional, defaults to `new Date()` if not included
     changefreq: 'weekly', // optional
     priority: 0.5 // optional
    }
  }
}
```

Next would already be aware of each route's `path`, so could automatically generate the necessary `loc` field. If deployed to Vercel, the `loc`'s `origin` could be automatically supplied via `process.env.VERCEL_URL`. Otherwise, users would need some mechanism to provide an `origin`, likely via `next.config.js`.

For SSR pages, `changefreq` could automatically default to `always`.

To exclude a SSG/SSR page from the sitemap, users could return `sitemap: false`.

**Comment by filipesmedeiros** [comment]
@natemoo-re I like that option, it seems to make sense. Although it kinda feels like `getStaticProps()` (because of its name) shouldn't do anything besides getting the props to the page. Maybe a better place would be `getStaticPaths()`? This one is actually building the sitemap, quite literally.

Also, while this is not implemented into next, do you guys know of an automated solution that scrapes the build output and generates the file? Thanks!
